wn/_core.py

```diff
@@ -1546,6 +1547,117 @@ def words(
+        data: Literal[True] = True
+    ) -> list[Form]: ...
+
+    # fallback for non-literal bool argument
+    @overload
+    def lemmas(
+        self,
+        form: Optional[str] = None,
+        pos: Optional[str] = None,
+        *,
+        data: bool
+    ) -> Union[list[str], list[Form]]: ...
+
+    def lemmas(
+        self,
+        form: Optional[str] = None,
+        pos: Optional[str] = None,
+        *,
+        data: bool = False
+    ) -> Union[list[str], list[Form]]:
+        """Return the list of lemmas for matching words in this wordnet.
+
+        Without any arguments, this function returns all lemmas in the
+        wordnet's selected lexicons. A *form* argument restricts the
+        words to those matching the given word form, and *pos*
+        restricts words by their part of speech.
+
+        If the *data* argument is :python:`False` (the default), the
+        lemmas are returned as :class:`str` types. If it is
+        :python:`True`, :class:`wn.Form` objects are used instead.
+
+        Example:
+
+            >>> wn.Wordnet().lemmas('wolves')
+            ['wolf']
+            >>> wn.Wordnet().lemmas('wolves', data=True)
+            [Form(value='wolf')]
+
+        """
+        kwargs: dict = {
+            'lexicons': self._lexconf.lexicons,
+            'search_all_forms': self._search_all_forms,
+        }
+
+        # if there's no form, query all lemmas
+        if form is None:
+            lemma_data = list(find_lemmas(pos=pos, **kwargs))
+            if data:
+                return [_make_form(*lemma) for lemma in lemma_data]
+            else:
+                return [lemma[0] for lemma in lemma_data]
+
+        # if there's a form, we may need to lemmatize and normalize
+        lemmatize = self.lemmatizer
+        normalize = self._normalizer
+        kwargs['normalized'] = bool(normalize)
+
+        forms = lemmatize(form, pos) if lemmatize else {}
+        # if no lemmatizer or word not covered by lemmatizer, back off to
+        # the original form and pos
+        if not forms:
+            forms = {pos: {form}}
+
+        # collect all lemmas in one go
+        all_lemma_data = [
+            lemma
+            for _pos, _forms in forms.items()
+            for lemma in find_lemmas(forms=_forms, pos=_pos, **kwargs)
+        ]
+
+        if not all_lemma_data and normalize:
+            all_lemma_data = [
+                lemma
+                for _pos, _forms in forms.items()
+                for lemma in find_lemmas(
+                    forms=[normalize(f) for f in _forms], pos=_pos, **kwargs
+                )
+            ]
+
+        # deduplicate lemmas while preserving order
+        if data:
+            unique_lemmas: list[Form] = []
+            seen: set[str] = set()
+            for lemma in all_lemma_data:
+                if lemma[0] not in seen:
+                    unique_lemmas.append(_make_form(*lemma))
+                    seen.add(lemma[0])
+            return unique_lemmas
```

Comment:
Deduping by the `str` lemma form probably makes sense when `data=False`, but when `data=True` I'd expect all lemmas to be returned. Consider "project", which is pronounced differently when it's a noun and when it's a verb.

For words, senses, and synsets, the `_find_helper()` dedupes based on the hash of the objects, which is based on the tuple of `(id, lexicon)`. Form IDs are optional (largely unused and always `None` for lemmas), so that doesn't work. Since there's a 1-1 correlation between lexical entries and lemmas, your `find_lemmas()` function could also return the lexical entry ID and then use that with the form's lexicon to dedupe (i.e., `hash((entry_id, form_lexicon))`).

Reply:
That makes sense